### PR TITLE
Add maxLineLength option to Grammar to only tokenize the first n characters of a line

### PR DESCRIPTION
**Pull request by wiggzz**
This will allow skipping tokenization of lines that are too long.  This is to prevent hangs on opening files with long lines, https://github.com/atom/atom/issues/979.

### CODE REVIEW

**Comment by Alhadis** [comment]
@wiggzz Just curious, how were you able to test this locally?

I'm trying to add a new property to grammars, and I can't seem to get my development copy of `first-mate` running. Have started Atom in dev mode, and it's even showing `console.log` calls on start-up... but anything called from within the Grammar class's constructor isn't showing up.

I'm guessing it's running the code asynchronously from Atom's bundled version... how did you manage to work around it? :)

**Comment by wiggzz** [comment]
You have to somehow shim your development version of first-mate into atom. So something like `npm install && npm pack` on first-mate, and then untar the resulting package directly into the node_modules/ directory of atom where first-mate was should do it.

**Comment by Alhadis** [comment]
Perfect, thank you! :D

Actually, I seem to be having troubles with the Oniguruma module... the rebuilt version of Atom is throwing an error on startup: *"Module version mismatch. Expected 47, got 46"*.

**Comment by wiggzz** [comment]
I've modified this pull request slightly so that it actually tokenizes the first `maxLineLength` characters of the line, since this will be a better experience. It still should bring performance improvements though since oniguruma will no longer be fed massive long lines of text that it could potentially spend a long time scanning.

File: src/grammar.coffee [hunk 1]
```diff
@@ -98,8 +98,9 @@ class Grammar
   # * `ruleStack` An {Array} of rules representing the tokenized state at the
   #   end of the line. These should be passed back into this method when
   #   tokenizing the next line in the file.
-  tokenizeLine: (line, ruleStack, firstLine=false, compatibilityMode=true) ->
+  tokenizeLine: (fullLine, ruleStack, firstLine=false, compatibilityMode=true) ->
```
Comment (jerone) [inline src/grammar.coffee]: Forgot to rename the variable in the docs above: https://github.com/atom/first-mate/pull/67/files#diff-30b5df382fb36749d0d1d282b2df9a53R82

**Comment by wiggzz** [comment]
Thanks @jerone. Fixed it.

**Comment by maxbrunsfeld** [comment]
Hey @wiggzz; I'm really sorry that we didn't respond to this PR before. Our open-source maintenance of this repo has been very poor.

Anyway, we finally got around to fixing this. Our solution was slightly different in that we reused the scope-closing logic in a way that didn't require allocating [a closure](https://github.com/atom/first-mate/pull/67/files#diff-30b5df382fb36749d0d1d282b2df9a53R129). But it's amusing how similar the code we wrote is to yours.

Apologies for not reusing this good code that you wrote.

**Comment by wiggzz** [comment]
No problem @maxbrunsfeld. It's hard to keep up with these things, just glad y'all are getting around to fixing it 😁

**Pull request closed without merging**